src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
+        return currentSetElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkPositionIndex(index);
+        Node<T> deletedElement = deletedLinks(null, index);
+        return deletedElement.value;
     }
 
     @Override
     public boolean remove(T object) {
+        int counter = 0;
+        Node<T> element = head;
+        while (element != null) {
+            if (element.value == object || element.value != null && element.value.equals(object)) {
+                deletedLinks(element, counter);
+                return true;
+            }
+            element = element.next;
+            counter++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private Node<T> next;
+        private T value;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.next = next;
+            this.value = value;
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index: "
+                    + index + " is out of bounds this list size: " + size);
+        }
+    }
+
+    private Node<T> findElementByIndex(int index) {
+        Node<T> nextElementByIndex;
+        if (index < size / 2) {
+            nextElementByIndex = head;
+            int counter = 0;
+            while (counter < index) {
+                nextElementByIndex = nextElementByIndex.next;
+                counter++;
+            }
+        } else {
+            nextElementByIndex = tail;
+            int counter = size - 1;
+            while (counter > index) {
+                nextElementByIndex = nextElementByIndex.prev;
+                counter--;
+            }
+        }
+        return nextElementByIndex;
+    }
+
+    private Node<T> deletedLinks(Node<T> deletedLinksNode, int index) {
+        Node<T> deletedNode = deletedLinksNodeByIndex(index);
+        if (deletedNode == null) {
+            if (deletedLinksNode == null) {
+                deletedNode = findElementByIndex(index);
+            } else {
+                deletedNode = deletedLinksNode;
+            }
+            deletedNode.next.prev = deletedNode.prev;
+            deletedNode.prev.next = deletedNode.next;
+        }
+        size--;
+        return deletedNode;
+    }
+
+    private Node<T> deletedLinksNodeByIndex(int index) {
+        Node<T> deletedElement;
+        if (size == 1) {
+            return head;
+        }
+        if (index == size - 1) {
+            deletedElement = tail;
+            tail = tail.prev;
+            tail.next = null;
+            return deletedElement;
+        }
+        if (index == 0) {
```

Comment:
could we just implement one method, that delete node that we pass as incoming parameter?
`private T unlink(Node<T> node)`